sdks/go/pkg/beam/core/runtime/exec/translate.go

```diff
@@ -372,32 +386,43 @@ func (b *builder) makeLink(from string, id linkID) (Node, error) {
 
 			switch op {
 			case graph.ParDo:
-				n := &ParDo{UID: b.idgen.New(), Inbound: in, Out: out}
-				n.Fn, err = graph.AsDoFn(fn, graph.MainUnknown)
+				dofn, err := graph.AsDoFn(fn, graph.MainUnknown)
 				if err != nil {
 					return nil, err
 				}
-				n.PID = transform.GetUniqueName()
-
-				input := unmarshalKeyedValues(transform.GetInputs())
-				for i := 1; i < len(input); i++ {
-					// TODO(herohde) 8/8/2018: handle different windows, view_fn and window_mapping_fn.
-					// For now, assume we don't need any information in the pardo payload.
-
-					ec, wc, err := b.makeCoderForPCollection(input[i])
-					if err != nil {
-						return nil, err
+				switch urn {
+				case urnPairWithRestriction:
+					u = &PairWithRestriction{UID: b.idgen.New(), Fn: dofn, Out: out[0]}
+				case urnSplitAndSizeRestrictions:
+					u = &SplitAndSizeRestrictions{UID: b.idgen.New(), Fn: dofn, Out: out[0]}
+				default:
+					n := &ParDo{UID: b.idgen.New(), Fn: dofn, Inbound: in, Out: out}
+					n.PID = transform.GetUniqueName()
+
+					input := unmarshalKeyedValues(transform.GetInputs())
+					for i := 1; i < len(input); i++ {
+						// TODO(herohde) 8/8/2018: handle different windows, view_fn and window_mapping_fn.
```

Comment:
NoActionRequiredNit: I don't like that this gets to 6 indentations deep, which is a bit of a readability concern, but I'm not certain that moving this to a helper function would be that helpful. We'd probably need to break it into two helpers, one from handling the graphx.URNDoFn, and then this graph.ParDo block. 

 These are a changes for another PR though.

Reply:
Agreed. I was iffy on it myself since even 5 indentations deep was already getting difficult to follow.